### DISCUSSION TITLE
3주차 Discussion

### DISCUSSION THREAD

**Discussion by chloe-codes1** [slippStudy/java-network-study#7] [General]
## 📢 책을 읽으면서 궁금한 점들을 자유롭게 적고 대답해 주세요

3주차 진행 방식

> 진행자: 정규님
> 

- 5 ~ 6장 읽기
- Discussion 에 질문 or 인상 깊은 점 2개 적어오기

**Comment by jinwookh** [comment]
1. byteStream을 사용했을 때 못 읽는 문자가 있을까요? 만약 없다면 문자스트림으로 읽는 모든 것은 byteStream으로도 읽을 수 있다고 봐도 되겠죠?
또 binary 파일을 문자스트림으로 읽고 문자스트림으로 썼을 때, 복사가 잘 되는지도 궁금합니다.

3. 버퍼 512byte를 만들어서 사용했을 때와, bufferStream을 사용했을 때 두 개의 성능 차이를 비교하고 싶습니다.

**Reply by LenKIM** [reply]
이미지(.jpg)(비정형된) ----> byteStreamOutstrea 직렬화 ---------> 역직렬화 -------> (형변환) 이미지(.jpg)

Reader  // Writer 용도에 맞는 사용인지?

**Comment by kangbada0728** [comment]
인상깊었던 점

1. 평소에 쓸때는 System.out.prinln 과 같이 쓰는 것만 썼는데, 해당 장을 볼수록 입력과 출력을 하는 방법은 정말 다양하다는 것을 느낌

궁금한 점

1. 키보드로 글자를 입력하면 OS 키보드 버퍼에 입력한 글자가 쌓이고, 다시 자바에서 Bufferer 를 사용하면 다시 또 자바 Buffer 에 글자가 쌓인다. OS 까지 생각해보면 버퍼를 이중으로 사용하고 있는건데 문제는 없을까?

**Comment by yangyunhyeuk** [comment]
[해당 챕터를 읽으며 인상 깊었던 점] 
마샬링이란 데이터를 특정 통신 채널로 보낼 수 있는 형태로 바꾸는 것으로, 
객체의 경우 마샬링을 하려면 Serializable 인터페이스를 구현해야 한다. 

그 과정에 있어서 transient란 예약어를 통해 마샬링되기를 원치않는 필드에 대한 처리도 할 수 있다. 
이처럼 직렬화하는 과정에 있어서 단순히 직렬화, 역직렬화하려는 목적에 의미를 두기보다는 
데이터 통신에 있어서 유의미한 데이터를 만드는 일련의 과정으로 느껴서, 이 점이 새롭게 다가왔다. 


[궁금한 점] 
Serializable이란 키워드는 몇 번 본 적이 있는데, transient 키워드는 실제 사용하는 케이스를 못봤다.
혹시 해당 키워드를 사용한 경험이 있는지, 있다면 어느 상황에서 사용했는지가 궁금합니다.

**Reply by chloe-codes1** [reply]
ref.
[The transient Keyword in Java](https://www.baeldung.com/java-transient-keyword)

**Comment by LenKIM** [comment]
- JSON, XML 등의 직렬화/역직렬화가 방식이 있음에도 ObjectStream 이 있다. 왜 직렬화를 사용하지 않고, JSON, XML 와 같은 것을 사용하는가?
        - 성능 차이가 얼마나 나나?
 
- 그럼 java.io.Serializable 인터페이스는 어떤 동작을 하는가 ?

**Reply by LenKIM** [reply]
**인상깊었던 점**

이제 System.out.println 코드를 이해할 수 있다.

**Reply by LenKIM** [reply]
패키지가 중요하다. 접근제어자 중요하다.

**Comment by chloe-codes1** [comment]
### 인상 깊었던 점

- BufferedWriter 에 저장된 데이터를 flush() 하거나 close() 하지 않으면 프로그램이 종료될 때 까지 해당 memory를 차지하고 있을 것이므로 리소스를 안전하게 해지하도록 유의해야겠다

### 궁금한 점

- Java에서 marshalling 과 unmarshalling 을 할 수 있는 가장 효율적인 방법들에는 무엇이 있을까?

**Reply by chloe-codes1** [reply]
ref. 
- [Is Protocol Buffers (ProtoBuf) really lighter and faster compared to JSON?](https://medium.com/@kn2414e/is-protocol-buffers-protobuf-really-lighter-and-faster-compared-to-json-681c6bee5d93)
- [마샬링(Marshalling) vs 직렬화(Serialization)](https://hyesun03.github.io/2019/09/08/marshalling-vs-serialization/)

**Reply by LenKIM** [reply]
성능 관점? Jackson? 상위호환성과 하위호환성?

binary, json
